### PR TITLE
Enable support for custom filesystem

### LINKED ISSUE #116: New FS not supported by llama-parse
llama-index now allow us to use Python FS (`AbstractFileSystem`) to load files so they don't need to exist in the local disk.

I couldn't find a way to make it work with llama-parse because it expects a path to a local file.

Are there any plans to make it possible?

### Error

`LlamaParse.load_data() got an unexpected keyword argument 'fs'. Skipping...`

### PR DESCRIPTION
**Pull request by thiagosalvatore**
Custom FileSystems that are now available on llama-index were not available on llama-parse. We can now use it to download files using any reader like S3Reader.

Fixes #116

### CODE REVIEW

**Comment by thiagosalvatore** [comment]
@logan-markewich can you take a look at this when you have some time, please?

File: llama_parse/base.py [hunk 1]
```diff
@@ -214,34 +219,37 @@ def validate_api_key(cls, v: str) -> str:
...
     def validate_base_url(cls, v: str) -> str:
         """Validate the base URL."""
         url = os.getenv("LLAMA_CLOUD_BASE_URL", None)
         return url or v or DEFAULT_BASE_URL
 
     # upload a document and get back a job_id
-    async def _create_job(self, file_path: str, extra_info: Optional[dict] = None) -> str:
-        file_path = str(file_path)
-        file_ext = os.path.splitext(file_path)[1]
+    async def _create_job(self, file_path: str | PurePath, extra_info: Optional[dict] = None, fs: Optional[AbstractFileSystem] = None,) -> str:
```
Comment (logan-markewich) [inline llama_parse/base.py:232]: Should be using `Union` and `Dict` typing, to support python3.8

File: llama_parse/base.py [hunk 2]
```diff
@@ -254,6 +262,12 @@ async def _create_job(self, file_path: str, extra_info: Optional[dict] = None) -
         job_id = response.json()["id"]
         return job_id
 
+    @staticmethod
+    def __get_filename(f: TextIOWrapper | AbstractBufferedFile) -> str:
```
Comment (logan-markewich) [inline llama_parse/base.py:266]: Any reason for this to be a static method? Or have a double underscore name? (also, should use union typing here again)

File: llama_parse/base.py [hunk 3]
```diff
@@ -214,34 +219,37 @@ def validate_api_key(cls, v: str) -> str:
...
         headers = {"Authorization": f"Bearer {self.api_key}"}
 
         # load data, set the mime type
-        with open(file_path, "rb") as f:
-            mime_type = mimetypes.guess_type(file_path)[0]
-            files = {"file": (f.name, f, mime_type)}
+        fs = fs or get_default_fs()
+        with fs.open(file_path, "rb") as f:
+            mime_type = mimetypes.guess_type(str_file_path)[0]
+            files = {"file": (self.__get_filename(f), f, mime_type)}
```
Comment (logan-markewich) [inline llama_parse/base.py:252]: I don't think this works, at least locally, it doesn't work for me (also breaks all the tests)

Comment (thiagosalvatore) [reply llama_parse/base.py:257]: I am using it in production right now and it is working properly. What is the error that you are getting locally?

The issue with the path is that I've seen situations in production where people upload a file with `.txt` but the file is actually a `.csv` and it doesn't work.

**Comment by sheresaidon** [comment]
@logan-markewich whenever you have a chance could you please review, currently llama parse is giving me the fs error when trying to use it with GCS Reader

parser = LlamaParse(result_type="markdown", api_key=LLAMA_PARSE_API_KEY,verbose=True, show_progress=True)

    reader = GCSReader(
        bucket=bucket_name,
        key=file_path,  # Use the extracted file path
        file_extractor={".pdf": parser, ".docx": parser, ".txt": parser},
        service_account_key=service_account_key_dict
    )


LlamaParse.load_data() got an unexpected keyword argument 'fs'. Skipping...

**Comment by thiagosalvatore** [comment]
@logan-markewich any updates here? I've been using this fix in prod for a while without any issues

**Comment by gjyoungjr** [comment]
any timeline on when this PR will get merged? my team is also blocked by not having this featue. 

does anyone know a current work around to load files that are not local in your directory?

**Review by logan-markewich — APPROVED** [review]

**Comment by logan-markewich** [comment]
:sigh: this still doesn't work guys lol

```
>>> from llama_parse import LlamaParse
>>> documents = LlamaParse(api_key="llx-jwAQZL8T38onyL9hKBOXyRtnuCU0Fk3z7tmDhIT3L0GEfohJ").load_data("2023.acl-srw.0.pdf")
Error while parsing the file '2023.acl-srw.0.pdf': '_io.BufferedReader' object has no attribute 'full_name'
>>> 
```

File: llama_parse/base.py (hunk 3 above)
Comment (logan-markewich) [reply llama_parse/base.py:175]: :sigh: this still doesn't work guys lol

```
>>> from llama_parse import LlamaParse
>>> documents = LlamaParse(api_key="llx-...").load_data("2023.acl-srw.0.pdf")
Error while parsing the file '2023.acl-srw.0.pdf': '_io.BufferedReader' object has no attribute 'full_name'
>>> 
```

**Review by logan-markewich — CHANGES_REQUESTED** [review]
This is still broken actually, see above

**Comment by logan-markewich** [comment]
Ok, I think this works, but haven't actually tested it with an external file system

**Comment by rohanprashant12** [comment]
Thanks @logan-markewich and @thiagosalvatore!

**Pull request merged**